src/main/java/com/pylon/spokestack/microsoft/BingSpeechClient.java

```diff
@@ -0,0 +1,375 @@
+            .putShort((short) 16)                   // bits per sample
+            .put("data".getBytes("US-ASCII"))       // begin data chunk
+            .putInt(0);                             // data chunk size
+        sendAudio(message);
+        flush();
+    }
+
+    /**
+     * transmits an audio frame over the websocket.
+     * @param frame the audio frame buffer to send
+     * @throws Exception on error
+     */
+    public void sendAudio(ByteBuffer frame) throws Exception {
+        if (this.socket == null)
+            throw new IllegalStateException();
+
+        if (frame == null || this.buffer.remaining() < frame.capacity())
+            flush();
+
+        if (this.buffer.position() == 0) {
+            byte[] header = (
+                "Path:audio\r\n"
+                + String.format("X-RequestId:%s\r\n", this.requestId)
+                + String.format("X-Timestamp:%s\r\n", now())
+                + "Content-Type:audio/x-wav\r\n"
+            ).getBytes("US-ASCII");
+
+            this.buffer
+                .order(ByteOrder.BIG_ENDIAN)        // protocol endian
+                .putShort((short) header.length)    // message header length
+                .put(header);                       // message header
+        }
+
+        if (frame != null) {
+            frame.rewind();
+            this.buffer.put(frame);                 // message body
+        }
+    }
+
+    /**
+     * sends an empty audio frame indicating the end of the turn.
+     * @throws Exception on error
+     */
+    public void endAudio() throws Exception {
+        if (this.socket == null)
+            throw new IllegalStateException();
+
+        sendAudio(null);
+        flush();
+    }
+
+    private void flush() {
+        if (this.buffer.position() > 0) {
+            this.buffer.flip();
+            this.socket.send(ByteString.of(this.buffer));
+            this.buffer.clear();
+        }
+    }
+
+    private String now() {
+        SimpleDateFormat formatter =
+            new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'000Z'");
+        formatter.setTimeZone(TimeZone.getTimeZone("UTC"));
+        return formatter.format(new Date());
+    }
+
+    /** okhttp socket callback. */
+    private class SocketListener extends WebSocketListener {
+        private final Gson gson = new Gson();
+        private final Listener listener = BingSpeechClient.this.listener;
+
+        @Override
+        public void onMessage(WebSocket s, String message) {
+            String requestHeader = String.format(
+                "X-RequestId:%s",
+                BingSpeechClient.this.requestId
+            );
+            if (message.contains(requestHeader))
+                if (message.contains("Path:speech.phrase"))
+                    onSpeechPhrase(message);
+        }
+
+        @Override
+        public void onClosed(WebSocket s, int code, String reason) {
+            if (code != 1000)
```

Comment:
they love their magic numbers

Reply:
in this case, it's a websocket standard: https://tools.ietf.org/html/rfc6455#section-7.4.1